### ISSUE TITLE
Cannot build when NODE_ENV is set to production

### ISSUE THREAD

**Issue by FREEZX** [meanjs/mean#360]
if NODE_ENV process variable is set to production, the following line loads production.js when trying to build:
https://github.com/meanjs/mean/blob/master/config/config.js#L14
This results in the build process trying to build and minify from the destination files, which fails.
To fix this, an env:development task should be added and called before loadConfig in the build task.

**Comment by lirantal** [comment]
what does grunt have anything to do with the config setting? there's not a bug in the config right?
anyway I didn't really understand what's wrong, or what's failing.

**Comment by FREEZX** [comment]
You run grunt build with your NODE_ENV set to production.
It calls loadConfig, which loads production config, with css files array pointing to public/dist/application.min.css
cssmin tries to minify the non-existent file into itself.
Fail.

**Comment by ilanbiala** [comment]
@FREEZX did you delete it or something? It is in this repo...https://github.com/meanjs/mean/tree/master/public/dist

**Comment by FREEZX** [comment]
I did delete it but it's silly to minify a css onto itself, without getting all the other css files.

**Comment by FREEZX** [comment]
This is what happens:
assume the environment is set to production.
The grunt build command does the following:

loadConfig - loads the config for production - meaning that in config.assets.css there will only be the minified css file (application.min.css), and inherently applicationCSSFiles will take the same value.

cssmin - minifies all files together defined in applicationCSSFiles and puts the minified version inside application.min.css (in this case it's application.min.css minifying into itself - not taking into account all the real css files located in public/modules/*/css)

**Comment by ilanbiala** [comment]
@FREEZX that file is a concatenation and minification of all the CSS in the modules...

@lirantal @NeverOddOrEven what do you guys think about having the build task create 1 production CSS file with all the CSS from libraries and modules in it? I think that makes sense because it reduces requests.

**Comment by FREEZX** [comment]
@ilanbiala It's not a concatenation and minification in this case is what i'm saying. An env:development task should be added in the beginning of the build for it to work properly.

**Comment by lirantal** [comment]
@ilanbiala sounds interesting but it requires some thinking, and also how to manage the CSS aggregation. Some CSS may need to be added before another, how do we get that worked out?

**Comment by FREEZX** [comment]
@lirantal Order of css statements is not important AFAIK.

**Comment by ilanbiala** [comment]
@lirantal I would do all libs first, then our CSS. I would order the libs corresponding to how the JS is ordered.

**Comment by Edo78** [comment]
I really need this issue to be solved. There's a ETA for its solution or can I help somehow?

BTW @FREEZX order of css statements IS important. This is why they are name **Cascading** Style Sheets

**Comment by Edo78** [comment]
A few hours ago I wrote a comment on issue #183 but now it looks the two issues could be related ...
BTW I'm copying here my comment.

I just generate a new app `yo meanjs` then I tested that `NODE_ENV=development grunt` works as intended but `NODE_ENV=production grunt`  give no errors but the app isn't showing anymore

`NODE_EVN=production grunt build` generate public/dist/application.js and public/dist/application.min.js both empty (and no min.css at all)

**Comment by FREEZX** [comment]
@Edo78 That's because applicationCSSFiles in the case of production environment is the minified css file, and if it's empty it gives back an empty output. Else it just copies itself on top of itself, without taking any of your changes into consideration.
If production is set to development, it takes all the non-minified css files (Because that's what's in applicationCSSFiles in this case) and properly outputs the minified css with your changes.

tl;dr: never build with production environment, add a first step in your build process that sets the environment to development temporarily (just like in the test task).

**Comment by Edo78** [comment]
@FREEZX ok, my bad for the build in production and thanks for your solution

**Comment by mbukosky** [comment]
They everyone, I lost a few hours today with this issue. I was finally able to get it working with @FREEZX  comment. I ran into this issue because of a deployment requirement with heroku. Heroku needs to have a grunt task called <i>heroku:production</i> for a production deployment. I had the same issue of having a blank landing page.

I was finally able to get it working by adding the following env:development code to my grunt file.

```javascript
...
    env: {
      development: {
        NODE_ENV: 'development'
      },
      test: {
        NODE_ENV: 'test'
      },
      secure: {
        NODE_ENV: 'secure'
      }
    },
...

grunt.registerTask('build', ['env:development', 'lint', 'loadConfig', 'ngAnnotate', 'uglify', 'cssmin']);

grunt.registerTask('heroku:production', ['build']);

```

I hope this helps out someone else.

**Comment by davestacey** [comment]
I tried the solutions suggested in "Figure out why NODE_ENV = 'production' doesn't work", but for some reason the grunt.registerTask('build', ['env:development', 'lint', 'loadConfig', 'ngAnnotate', 'uglify', 'cssmin' ]); never generates the public/dist/application.min.js. Since I never get public/dist/application.js, then uglify "files: {'public/dist/application.min.js': 'public/dist/application.js'} never happens. 

I can bypass this issue by changing env:production.js to match all.js in the section related to assets: { js... and put the paths to the js themselves but this defeats the concat and uglify steps. 

Where is the step that writes public/dist/application.js ? (not min.js ). I looked into the getGlobbedFiles in config.js but can't find the step where we write out application.js to public/dist. I can change gruntfile.js and make this happen but I want to use the correct solution.

I'm watching #183 "Production CSS/JS minification does not glob all source files"  for a solution
and I also tried using gruntfile.js https://github.com/mbukosky/SpotifyUnchained/blob/72e08b67aeaf35a4bb4e9d2206d3d0bd8674c2a1/gruntfile.js

**Comment by FREEZX** [comment]
@davestacey ngAnnotate takes all js files defined in the config file and annotates them them into public/dist/application.js

**Comment by davestacey** [comment]
Thanks FREEZX, that's what I thought. I created with yo meanjs, I used the default gruntfile.js and also tried the gruntfile.js from "14- Fixing 'mean' build issue". When I deploy to Heroku, or run NODE_ENV=production grunt, it doesn't seem to be generating public/dist/application.js. Not a show stopper since I have a work around. To recreate the issue, I followed the process http://meanjs.org/generator.html. yo meanjs, grunt, I tried NODE_ENV=production grunt, that didn't create dist folder with js or css. Everything runs fine in ENV dev default mode http://localhost:3000/#!/.  Another question, the package.json yo creates has"version": "0.0.1", but via git clone is "version": "0.3.3", is yo creating mean.js v 0.0.1? I'm installing with yo since I like the CRUD module generator. Thanks for the help.


I also tried with gruntfile.js from https://github.com/mbukosky/SpotifyUnchained/blob/72e08b67aeaf35a4bb4e9d2206d3d0bd8674c2a1/gruntfile.js

**Comment by FREEZX** [comment]
@davestacey the generator generates with 0.0.1, because it's the initial state of your own application.
That happens here:
https://github.com/meanjs/generator-meanjs/blob/master/app/templates/_package.json#L4
From what i see, the github repository is pretty much up to date, so you should not be worried.

**Comment by davestacey** [comment]
Just to clarify, should I be able to run "NODE_ENV=production grunt" and have public/dist/application.js generated?

**Comment by FREEZX** [comment]
@davestacey Nope, only `grunt build` should create `public/dist/application.js` as well as `public/dist/application.min.js`

**Comment by davestacey** [comment]
ok I found what I was missing, I need to run "NODE_EVN=production grunt build", Edo78 also mentioned that earlier and I missed it, thanks for all the help!!

**Comment by mbukosky** [comment]
@davestacey I normally test my local production build using

```
NODE_ENV=production grunt heroku:production
```

Hope that helps

**Comment by ilanbiala** [comment]
If anyone wants to submit a PR that fixes this, that would be much appreciated.

**Comment by FREEZX** [comment]
Here
https://github.com/meanjs/mean/pull/464

**Comment by davestacey** [comment]
@mbukosky - That helped, sorry if I missed it in the documentation. Running "NODE_ENV=production grunt heroku:production" did the trick and everything is pushing to Heroku as expected. Thanks again.

**Comment by ilanbiala** [comment]
already fixed in 0.4.0.

**Comment by jloveland** [comment]
on the 0.4.0 branch, I got this working by running `grunt prod`

**Comment by simsolTech** [comment]
It runs if I use the contents of config/assets/default.js in config/assets/production.js

**Comment by Agent5** [comment]
Grunt build never works for me no matter the environment. I get an error stating ">> error: couldn't process source due to parse error
>> Unexpected token (1074:0)"

Has anyone else come across this?
